infrastructure/comms/src/connection/peer_connection/worker.rs

```diff
@@ -0,0 +1,320 @@
+                    _ => PeerConnectionState::Failed(PeerConnectionError::UnexpectedConnectionError),
+                };
+            },
+        }
+
+        Ok(())
+    }
+
+    /// The main loop for the worker. This is where the work is done.
+    /// The required connections are set up and messages processed.
+    fn main_loop(&mut self) -> Result<()> {
+        let monitor = self.connect_monitor()?;
+        let peer_conn = self.establish_peer_connection()?;
+        let consumer = self.establish_consumer_connection()?;
+
+        loop {
+            if let Some(msg) = self.receive_control_msg()? {
+                match msg {
+                    ControlMessage::Shutdown => break Ok(()),
+                    ControlMessage::SendMsg(frames) => {
+                        let payload = self.create_payload(frames)?;
+                        peer_conn.send(payload)?;
+                    },
+                    ControlMessage::Pause => {
+                        self.paused = true;
+                    },
+                    ControlMessage::Resume => {
+                        self.paused = false;
+                    },
+                }
+            }
+
+            if !self.paused {
+                self.forward_frames(&peer_conn, &consumer)?;
+            }
+
+            if let Ok(event) = monitor.read(1) {
+                self.handle_socket_event(event)?;
+            }
+        }
+    }
+
+    /// Handles socket events from the ConnectionMonitor. Updating connection
+    /// state as necessary.
+    fn handle_socket_event(&mut self, event: SocketEvent) -> Result<()> {
+        use SocketEventType::*;
+
+        match event.event_type {
+            Disconnected => {
+                let mut lock = acquire_write_lock!(self.connection_state)?;
+                *lock = PeerConnectionState::Disconnected;
+            },
+            Listening | Accepted | Connected => {
+                self.retry_count = 0;
+                let mut lock = acquire_write_lock!(self.connection_state)?;
+                match *lock {
+                    PeerConnectionState::Connecting(ref thread_ctl) => {
+                        *lock = PeerConnectionState::Connected(thread_ctl.clone());
+                    },
+                    PeerConnectionState::Connected(_) => {},
+                    ref s => {
+                        return Err(PeerConnectionError::StateError(format!(
+                            "Unable to transition to connected state from state '{}'",
+                            s
+                        ))
+                        .into());
+                    },
+                }
+            },
+            BindFailed | AcceptFailed | HandshakeFailedNoDetail | HandshakeFailedProtocol | HandshakeFailedAuth => {
+                let mut lock = acquire_write_lock!(self.connection_state)?;
+                *lock = PeerConnectionState::Failed(PeerConnectionError::ConnectFailed);
+            },
+            ConnectRetried => {
```

Comment:
Is this somehow followed by another `Listening | Accepted | Connected` event? Else how will we ever actually reach the connected state?

Reply:
Events are queued and read off in the order they happen (at least that seems to be the case, I think you'd have to inspect zero MQ's code to be sure). `ConnectRetried` happens when the connection cannot be made, and zero MQ is retrying for us, however I we need to check if we're in a `Connecting` state because if this is in any other state, no state transition should occur (as `Disconnect` should eventually be triggered if the connection drops) - Thanks!

Reply:
To answer your question, `ConnectRetried` is followed by `Connected` once a connection is made and that will transition this to a `Connected` state. `Accepted` is when someone connects to us (our connection is `Inbound`) and so probably doesn't trigger retries. As mentioned before, I'll remove `Listening` from the above.

Reply:
Ok, thanks